src/Workspaces/Core/Portable/CodeRefactorings/FixAllOccurences/DocumentBasedFixAllProvider.cs

```diff
@@ -0,0 +1,132 @@
+        {
+            _supportedFixAllScopes = supportedFixAllScopes;
+        }
+
+        /// <summary>
+        /// Produce a suitable title for the fix-all <see cref="CodeAction"/> this type creates in <see
+        /// cref="GetFixAsync(FixAllContext)"/>.  Override this if customizing that title is desired.
+        /// </summary>
+        protected virtual string GetFixAllTitle(FixAllContext fixAllContext)
+            => fixAllContext.GetDefaultFixAllTitle();
+
+        /// <summary>
+        /// Apply fix all operation for the code refactoring in the <see cref="FixAllContext.Document"/>
+        /// for the given <paramref name="fixAllContext"/>.  The document returned will only be examined for its content
+        /// (e.g. it's <see cref="SyntaxTree"/> or <see cref="SourceText"/>.  No other aspects of document (like it's properties),
+        /// or changes to the <see cref="Project"/> or <see cref="Solution"/> it points at will be considered.
+        /// </summary>
+        /// <param name="fixAllContext">The context for the Fix All operation.</param>
+        /// <param name="document">The document to fix.</param>
+        /// <param name="fixAllSpans">The spans to fix in the document.</param>
+        /// <returns>
+        /// <para>The new <see cref="Document"/> representing the content fixed document.</para>
+        /// <para>-or-</para>
+        /// <para><see langword="null"/>, if no changes were made to the document.</para>
+        /// </returns>
+        protected abstract Task<Document?> FixAllAsync(FixAllContext fixAllContext, Document document, ImmutableArray<TextSpan> fixAllSpans);
+
+        public sealed override IEnumerable<FixAllScope> GetSupportedFixAllScopes()
+            => _supportedFixAllScopes;
+
+        public sealed override Task<CodeAction?> GetFixAsync(FixAllContext fixAllContext)
+            => DefaultFixAllProviderHelpers.GetFixAsync(
+                fixAllContext.GetDefaultFixAllTitle(), fixAllContext, FixAllContextsHelperAsync);
+
+        private Task<Solution?> FixAllContextsHelperAsync(FixAllContext originalFixAllContext, ImmutableArray<FixAllContext> fixAllContexts)
+            => DocumentBasedFixAllProviderHelpers.FixAllContextsAsync(originalFixAllContext, fixAllContexts,
+                    originalFixAllContext.ProgressTracker,
+                    this.GetFixAllTitle(originalFixAllContext),
+                    GetFixedDocumentsAsync);
+
+        /// <summary>
+        /// Attempts to apply fix all operations returning, for each updated document, either
+        /// the new syntax root for that document or its new text.  Syntax roots are returned for documents that support
+        /// them, and are used to perform a final cleanup pass for formatting/simplication/etc.  Text is returned for
+        /// documents that don't support syntax.
+        /// </summary>
+        private async Task<Dictionary<DocumentId, (SyntaxNode? node, SourceText? text)>> GetFixedDocumentsAsync(
```

Comment:
This is the only method in `DocumentBasedFixAllProvider` types that is not shared between the one for code fixes and refactorings. There are quite a few differences that sharing it did not make sense. Existing code: https://github.com/dotnet/roslyn/blob/c9d9e122cfff99ce9508a54897254d407b148f48/src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DocumentBasedFixAllProvider.cs#L126-L169

Reply:
slightly surpirsed.  but that's ok if you did the due diligence to try to share and found it unhelpful.

Reply:
Yes, feel free to refactor/share this with a follow-up PR if you feel that it helps maintainability without hurting readability.